### ISSUE TITLE
Module system in NPM package for igv.js 2.0

### ISSUE THREAD

**Issue by eweitz** [igvteam/igv.js#701]
Following up on the newly available ES6 export (#674), I'd like to propose that the default [igv NPM package](https://www.npmjs.com/package/igv) prioritize support for `import` via ES6 as part of the upcoming igv.js 2.0 release.  

The current package supports `require` and not `import`, and it seems impossible to easily support non-standard and standard module systems in the same NPM package.

I see two options:

1. Support `import` in the default package, and provide a separate `igv-require` package
2. Support `require` in the default package , and provide a separate `igv-es6` package

Neither of those options would break anything, but they would optimize igv differently.

Option 1, making `import` the easier approach, would ensure that more new dependents (and more modern existing dependents) will be able to easily include the default package like they do their other dependencies.  The ES6 `import` standard has already replaced the non-standard `require` as the preferred way to include dependencies in all major JS frameworks and most modern JS libraries. That trend will only increase during the life of igv.js 2.x.

Option 2, making `require` the easier approach, would ensure that the handful of existing dependents using the igv NPM package can continue using the default package as they do now. Although some dependents use igv via NPM and `require`, given embedders I know and my GitHub searches of igv.js usage patterns, I suspect the vast majority include igv.js by hand-written build steps that pull it in by pointing to a URL at build- or run-time. And those few who get igv.js via NPM but do not migrate from the non-standard `require` syntax would still be supported by a separate `igv-require` package.

Either option would work, but I think the case for Option 1 is stronger.

**Comment by jrobinso** [comment]
To summarize,  igv.js will have 2 NPM packages,  one for use as an ES6 module (using "import"),  the second for all other uses  ("require", AMD, and plain script tag imports).   The question is,  which package gets the name "igv",  and which a qualifier.    BTW I'm not sure about "require" as the qualifier as its really everything other than an ES6 module, including "require" and plain script includes.

I don't see a big difference wrt ease-of-use, in both cases you have to find the correct package for the module system your are using.

@vrockai @renyaoxiang @paul-shannon @MariaNattestad  You have all mentioned NPM or require in issues posted here.   If you have an opinion here please post it here.   Also, if you use NPM please post a short note on how you are using it.   Note that the proposed change is for the upcoming 2.0 release.

**Comment by vrockai** [comment]
We're still using [Bower](https://bower.io/) for js packages in the project using the IGV browser. So far I had to use copy-pasted slightly edited IGV js file because of the lodash version clash AFAIR. We haven't yet switched to the `npm` package since the solution we got now works.

**Comment by jrobinso** [comment]
Thanks @vrockai -- the lodash problem was fixed long ago,  of course not in the released version (1.0.9).  In 2.0 there are no dependencies at all.

**Comment by paul-shannon** [comment]
I would be happy with either:

  require(“igv”)
or
  require(“igv.npm”)

or ….

And likely to move to ES6 in the not too distant future.

 - Paul

**Comment by eweitz** [comment]
> igv.js will have 2 NPM packages, one for use as an ES6 module (using "import"), the second for all other uses ("require", AMD, and plain script tag imports).

Supporting `import` and plain old `script` tags in the same NPM package is straightforward.  I put together [an example using traditional script tags with the experimental ES6 igv NPM package](https://jsfiddle.net/mq54x2y6/1/).

That's enabled by having both `igv.min.js` and `igv.es6.js` in the NPM package, which is an effect of the [package.json updates](https://github.com/igvteam/igv.js/issues/674#issuecomment-410520847) we discussed.

Looking closer, I think it would also be possible to support `require` in the same NPM package, using `igv.min.js` as in [your igv.js `require` example](http://igv.org/web/test/examples/igv-require.html).  My [attempt to do so with the `tmp_es6_igv` package](https://jsfiddle.net/mogbkej0/) failed, but that can likely be fixed with a small adjustment to igv.min.js.

I just found that my ideogram NPM package can work with `require` in addition to the existing `import` and `script` tag support, so the same should be possible with the igv NPM package.

In other words, having `"main": "dist/igv.esm.js"` in package.json for the default igv NPM package would enable ES6 `import` as expected by modern React, Angular and Vue apps, while still supporting traditional `script` tags and `require` in the same package.

**Comment by jrobinso** [comment]
@eweitz it would be awesome to have 1 NPM package.   Could you zip and attach a package.json to support that, or send a pull request?   You can assume I know nothing about NPM package.json and wouldn't be far off.

**Comment by eweitz** [comment]
Sounds good.  I'll fix that `require` issue, then send a pull request.

**Comment by eweitz** [comment]
Resolved per #702.

**Comment by jrobinso** [comment]
@eweitz I realize this is closed but its the right place to discuss NPM anyway.   I'm working on modifying the script to create the NPM package.   I don't want to include everything in the project or enumerate a zillion .npmignores, which will change over time and be error prone.   So I'm creating a tmp folder and copying what's needed to there.    For js I think it should be the minified files, to reduce the size.   Do you agree?   Should the map files be included?

So the minimal structure would be

package.json
LICENSE
README.md
dist/igv.esm.min.js
dist/igv.min.js

OTOH,  jquery includes everything.       Perhaps the entire dist folder?  Will that slow down importing of igv?

**Comment by eweitz** [comment]
@jrobinso, you could likely get by without using a Jenkins script to create the NPM package -- especially with the package.json improvements that were merged earlier this week.  

I just opened #706 to refine those package.json improvements.  That PR has more context and detail. 
 If it works for you, let's chat about this over there.
